Feat: 타이머 기능 개발

jungjinbeom opened the pull request:
### API 호출 커스텀 훅 개발
 - 스터디룸 정보 조회 API
 - 스터디룸 정보 업데이트 API

### 타이머 기능 개발 
 - 타이머 커스텀 훅 개발
 - 사운드 플레이 커스텀 훅 추가 
 - 타이머 종료 사운드 파일 추가
 
### 스터디 진행 기능 개발  
 - 스터디 상태 변경 함수 추가 
 - 스터디 상태 상수 파일 및 상수 값 추가 
 - 스터디 Step 타입 추가 
 - 스터디 상태 보관 및 변경 커스텀 훅 개발

### 스터디룸 레이아웃 적용 
- 스터디룸용 레이아웃 컴포넌트 적용

jungjinbeom committed:
feat : 타이머 기능 개발
- API 파일 및 로직 추가
- 스터디 진행 상수 파일 및 데이터 추가
- 타이머 기능 커스텀 훅 개발 중
- API 호출 커스텀 훅 개발

jungjinbeom committed:
feat : 타이머 기능 개발
- 스터디 진행 상수 파일 및 데이터 추가
- useAudio 커스텀 훅 개발
- useTImer 기능 커스텀 훅 개발
- 스터디룸 정보 조회, 업데이트 API 호출 커스텀 훅 개발
- 오디오 파일 추가
- studyRoom 레이아웃 변경

jungjinbeom committed:
fix: 사용하지 않는 컴포넌트 삭제, 파일 명 import 에러 수정
- main.tsx의 App.tsx import 시 파일 경로 에러 수정
- 랜딩페이지에서 사용되지 않는 Timer 컴포넌트 삭제

Seung-wan commented on src/apis/study-room/study-room-info.ts:
```diff
@@ -0,0 +1,16 @@
+  step: string;
+  participantCount: number;
+  updateAt: string;
+};
```
step 관련해서 상수 파일로 만들어주신게 있어서 이렇게 쓸 수 있는 상황이면 적용해주셔도 좋을 것 같네요~
```suggestion
type Response = {
  id: number;
  name: string;
  intro: string;
  step:Step;
  participantCount: number;
  updateAt: string;
};
```

Seung-wan commented on src/apis/study-room/study-room-info.ts:
```diff
@@ -0,0 +1,16 @@
+  updateAt: string;
+};
+
+export const studyRoomInfo = async ({ studyId }: Params): Promise<Response> => {
```
함수명에 http method도 들어가는게 좋지 않을까요?
함수명만 보면 동작이 예상이 안되는 것 같아요~

jungjinbeom committed:
fix: 오디오 파일 제거
- 중복되는 오디오 파일 제거

Seung-wan commented on src/apis/study-room/study-room-info.ts:
```diff
@@ -0,0 +1,16 @@
+
+export const studyRoomInfo = async ({ studyId }: Params): Promise<Response> => {
+  const response = await http.get(`/studyrooms/${studyId}`);
+  return response.data;
```
취향이지만 요렇게 표현해줘도 좋을 것 같네요~
```suggestion
  const { data } = await http.get(`/studyrooms/${studyId}`);
  
  return data;
```

Seung-wan commented on src/apis/study-room/study-room-info.ts:
파일명에도 개인적으로는 http method가 들어가주면 어떨까 싶네요!

Seung-wan commented on src/apis/study-room/study-step.ts:
```diff
@@ -0,0 +1,7 @@
+import { http } from '../api';
+
+type Params = { studyId: number; step: string };
+export const studyStep = async ({ studyId }: Params): Promise<void> => {
```
사소하지만 요런 개행도 챙겨주면 좋을 것 같습니다~
```suggestion
type Params = { studyId: number; step: string };

export const studyStep = async ({ studyId }: Params): Promise<void> => {
```

Seung-wan commented on src/constant/step.ts:
```diff
@@ -0,0 +1,7 @@
+  RETROSPECT: 'RETROSPECT',
+  RESTING: 'RESTING',
+  COMPLETED: 'COMPLETED',
+};
```
요건 어디서 사용되는 상수값인가요?

Seung-wan commented on src/pages/study-room/components/timer.tsx:
```diff
@@ -1,15 +1,47 @@
+  RETROSPECT: '회고',
+  RESTING: '휴식',
+  COMPLETED: '완료',
+};
```
constant에 step이 있어서 요것도 그쪽으로 모아두면 어떨까 싶네요~

Seung-wan replied:
이렇게 표현해줘도 좋을 것 같아요~
```suggestion
const STEP_TEXT: Record<string, string> = {
  PLANNING: '계획',
  STUDYING: '스터디',
  RETROSPECT: '회고',
  RESTING: '휴식',
  COMPLETED: '완료',
};
```

Seung-wan replied:
STEP_TEXT의 key값은 타입을 잡아줄 수 있을 것 같네요!

```ts
type StepTextKey = keyof typeof STEP;

const STEP_TEXT: Record<StepTextKey, string> = {
  PLANNING: '계획',
  STUDYING: '스터디',
  RETROSPECT: '회고',
  RESTING: '휴식',
  COMPLETED: '완료',
};
```

Seung-wan replied:
보통 요런 객체는 MAP이라는 이름으로도 많이 표현해주는 것 같습니다~
```suggestion
const STEP_LABEL_MAP: KeyType = {
  PLANNING: '계획',
  STUDYING: '스터디',
  RETROSPECT: '회고',
  RESTING: '휴식',
  COMPLETED: '완료',
};
```

Seung-wan commented on src/pages/study-room/components/timer.tsx:
```diff
@@ -1,15 +1,47 @@
+import { useMemo } from 'react';
+import useTimer from '../hook/useTimer';
+
+type KeyType = { [key in string]: string };
```
type에 -Type 접미사는 굳이 사용해주지 않아도 괜찮을 것 같아요~

Seung-wan commented on src/pages/study-room/components/timer.tsx:
```diff
@@ -1,15 +1,47 @@
+  const currentTime = useMemo(() => {
+    const min = Math.floor(seconds / 60);
+    const sec = seconds % 60;
+    document.title = `${String(min).padStart(2, '0')}:${String(sec).padStart(2, '0')}`;
```
요건 currentTime이라는 값을 구하는 useMemo 내부에 들어있는건 상당히 사이드이펙트처럼 느껴지네요~
별도로 useEffect에서 처리해주는게 좋을 것 같습니다!

Seung-wan commented on src/pages/study-room/components/timer.tsx:
```diff
@@ -1,15 +1,47 @@
+
+  const currentTime = useMemo(() => {
+    const min = Math.floor(seconds / 60);
+    const sec = seconds % 60;
```
개~~~인적으로는 요런식으로 표현해줘도 좋다고 생각이 들긴 하네요!
```suggestion
    const 분 = Math.floor(seconds / 60);
    const 초 = seconds % 60;
    
    return `${String(분).padStart(2, '0'):${String(초).padStart(2, '0')}`;
```

Seung-wan commented on src/pages/study-room/components/timer.tsx:
```diff
@@ -1,15 +1,47 @@
+  const { seconds, startTimer, status } = useTimer({
+    updateAt: '2024-03-11T14:55:32.513Z',
+    step: 'PLANNING',
+  });
```
format과 같은 option을 하나 더 만들어서 아예 useTimer 내부에서 화면에 보여줄 시간의 형식까지 맞춰주는것도 어떨까 싶네요~ 
```suggestion
  const { time, startTimer, status } = useTimer({
    updateAt: '2024-03-11T14:55:32.513Z',
    step: 'PLANNING',
    format: 'MM:SS'
  });
```

Seung-wan commented on src/pages/study-room/components/timer.tsx:
```diff
@@ -1,15 +1,47 @@
-        <Button size="small">시작</Button>
-        <Button size="small">리셋</Button>
-        <Button size="small">시간설정</Button>
+        <Button onClick={startTimer}>{`${STEP_TEXT[status]} 시작`}</Button>
```
button의 type은 default가 submit이어서 명시적으로 button으로 잡아줘도 좋을 것 같습니다~
```suggestion
        <Button type='button' onClick={startTimer}>{`${STEP_TEXT[status]} 시작`}</Button>
```

Seung-wan commented on src/pages/study-room/hook/useAudio.ts:
```diff
@@ -0,0 +1,10 @@
+import sound from '../../../assets/audio/flipdish-ringer.mp3';
+const useAudio = () => {
+  const audio = useRef(new Audio(sound));
+  const playSound = () => audio.current.play();
```
useCallback으로 래핑해줘도 좋을 것 같습니다~

Seung-wan replied:
이미 play라는 메소드가 audio에서는 익숙하기 때문에 그대로 play로 표현해주는건 어떨까요?
```suggestion
  const play = () => audio.current.play();
```

Seung-wan commented on src/pages/study-room/hook/useAudio.ts:
```diff
@@ -0,0 +1,10 @@
+import { useRef } from 'react';
+import sound from '../../../assets/audio/flipdish-ringer.mp3';
+const useAudio = () => {
```
요런것도 개행 넣어주시면 좋을 것 같습니다~
```suggestion
import sound from '../../../assets/audio/flipdish-ringer.mp3';

const useAudio = () => {
```

Seung-wan commented on src/pages/study-room/hook/useAudio.ts:
폴더명이 hook -> hooks가 되어도 좋을 것 같네요!

Seung-wan commented on src/pages/study-room/hook/useStepStatus.ts:
```diff
@@ -0,0 +1,10 @@
+
+const useStepStatus = (stepStatus: string) => {
+  const [status, setStatus] = useState(stepStatus);
+  const changeStepStatus = (step: string) => setStatus(step);
```
여기서 쓰이는 step도 타입으로 지정해주면 좋을 것 같네요~

Seung-wan commented on src/pages/study-room/hook/useStudyStepMutation.ts:
```diff
@@ -0,0 +1,11 @@
+  const studyStepMutation = useMutation({
+    mutationFn: studyStep,
+  });
+  return { studyStepMutation };
```
요렇게 표현해줘도 괜찮지 않을까요?
지금 구조에서는 사용하는 쪽의 인터페이스가 일반적인 기대와 다를 것 같아요~
```suggestion
  return useMutation({
    mutationFn: studyStep,
  });
```

Seung-wan commented on src/pages/study-room/hook/useTimer.ts:
```diff
@@ -0,0 +1,50 @@
+interface UseTimeProps {
+  updateAt: string;
+  step: string;
+}
```
요렇게 되어야 하지 않을까요?
```suggestion
interface UseTimerProps {
  updateAt: string;
  step: string;
}
```

Seung-wan replied:
컴포넌트가 아니기 때문에 꼭 -Props는 적용해주지 않아도 괜찮을 것 같네요!
```suggestion
interface UseTimer {
  updateAt: string;
  step: string;
}
```

Seung-wan commented on src/pages/study-room/hook/useTimer.ts:
```diff
@@ -0,0 +1,50 @@
+import { useRef, useState } from 'react';
+import useStepStatus from './useStepStatus';
+import { studyStep } from '../util/studyStep';
```
요건 함수 이름부터 util에 포함되진 않을 것 같네요~

Seung-wan commented on src/pages/study-room/hook/useStepStatus.ts:
```diff
@@ -0,0 +1,10 @@
+import { useState } from 'react';
+
+const useStepStatus = (stepStatus: string) => {
+  const [status, setStatus] = useState(stepStatus);
```
이렇게 표현되는게 네이밍이 일관적일 것 같네요~
```suggestion
const useStepStatus = (initialStepStatus: string) => {
  const [stepStatus, setStepStatus] = useState(initialStepStatus);
```

Seung-wan commented on src/pages/study-room/hook/useTimer.ts:
```diff
@@ -0,0 +1,50 @@
+}
+
+const DELAY = 1000;
+const STUDY_TIME = 1500;
```
STUDY_TIME은 외부에서 주입받아서 사용할 수 있어야 할 것 같아요~

Seung-wan commented on src/pages/study-room/hook/useTimer.ts:
```diff
@@ -0,0 +1,50 @@
+         * const data = studyStepMutation.mutate({ studyId: 1, step });
+         *
+         */
+        changeStepStatus(step);
```
step을 바꾸는 로직은 useTimer 외부에서 수행되는게 좋을 것 같네요~

Seung-wan commented on src/pages/study-room/hook/useTimer.ts:
```diff
@@ -0,0 +1,50 @@
+        setSeconds(STUDY_TIME);
+        return;
+      }
+      if (remainingSeconds > 0) {
```
조건이 2가지로 정확히 구분되서, `if (remainingSeconds > 0)` 조건을 early return 처리해주고, 다른 if문 하나는 제거해주면 좋을 것 같네요~

Seung-wan commented on src/pages/study-room/hook/useTimer.ts:
```diff
@@ -0,0 +1,50 @@
+         */
+        changeStepStatus(step);
+        clearInterval(intervalRef.current as NodeJS.Timeout);
+        timeRef.current = STUDY_TIME;
```
timeRef.current 값은 도중에 변하지 않는데, 여기서 STUDY_TIME값을 넣어주는 이유가 있나요?

Seung-wan commented on src/pages/study-room/hook/useTimer.ts:
```diff
@@ -0,0 +1,50 @@
+        changeStepStatus(step);
+        clearInterval(intervalRef.current as NodeJS.Timeout);
+        timeRef.current = STUDY_TIME;
+        setSeconds(STUDY_TIME);
```
타이머가 종료됐을때 00:01, 00:00 으로 보여졌다가 25:00이 될 것 같은데 이런 처리는 reset과 같은 함수에서 해주는게 맞지 않을까요?

Seung-wan commented on src/pages/study-room/util/studyStep.ts:
```diff
@@ -0,0 +1,16 @@
+export const studyStep = (type: string) => {
```
step으로 표현되는게 조금 더 이해하기 편할 것 같네요~
```suggestion
export const studyStep = (step: string) => {
```

Seung-wan replied:
함수의 동작을 보면 대략 이런식으로 표현되어야 하지 않을까요?
```suggestion
export const getNextStudyStep = (type: string) => {
```

Seung-wan replied:
해당 함수는 스터디룸의 스텝을 구하는 함수인데 유틸로 보기엔 어려울 것 같아요.
도메인 로직이 아닐까요?

Seung-wan commented on src/pages/study-room/util/studyStep.ts:
```diff
@@ -0,0 +1,16 @@
+export const studyStep = (type: string) => {
+  switch (type) {
+    case 'PLANNING':
+      return 'STUDYING';
+    case 'STUDYING':
+      return 'RETROSPECT';
+    case 'RETROSPECT':
+      return 'RESTING';
+    case 'RESTING':
+      return 'COMPLETED';
+    case 'COMPLETED':
+      return 'PLANNING';
```
정의해놓은 상수를 이용해서 표현해주는게 좋을 것 같아요~

junhyeong00 replied:
백엔드에서 스텝을 바꿔서 전달해주는데 프론트에서도 이런식으로 코드를 작성한 이유가 있으신가요?

jungjinbeom replied:
다음 스터디 정보를 업데이트 하기 위해서 코드를 작성했습니다.

Seung-wan commented on src/pages/study-room/util/studyStep.ts:
```diff
@@ -0,0 +1,16 @@
+export const studyStep = (type: string) => {
+  switch (type) {
+    case 'PLANNING':
+      return 'STUDYING';
+    case 'STUDYING':
+      return 'RETROSPECT';
+    case 'RETROSPECT':
+      return 'RESTING';
+    case 'RESTING':
+      return 'COMPLETED';
+    case 'COMPLETED':
+      return 'PLANNING';
+    default:
+      return 'ERROR';
+  }
```
Lookup Table로 표현해줘도 좋을 것 같네요~

kkkkkSE committed:
Chore: 스터디룸 목록 관련 테스트를 위한 fixture 데이터 추가

kkkkkSE committed:
Chore: Jest가 @(root 경로)를 인식할 수 있도록 설정 추가

kkkkkSE committed:
Feat: 스터디룸 목록 조회 기능 추가

kkkkkSE committed:
Refactor: 데이터 타입 지정 시점을 쿼리 -> api 호출 시점으로 변경

kkkkkSE committed:
Chore: apis/study-rooms -> apis/study-room 폴더명 변경

kkkkkSE committed:
Style: 코드리뷰를 반영한 단순 코드 스타일 변경

jungjinbeom committed:
chore : 리뷰 반영
- STEP_TEXT -> STEP_LABEL_MAP로 상수명 변경
- 스터디 진행 상황 함수 명 변경
   studyStep -> getNextStudyStep
- 스터디 진행 함수 Switch문에서 Lookup 테이블 형식으로 변경
- SignitureKey 타입 명 변경
  KeyType -> StepTextKey
- currentTime 내부 변수 명 변경
  min -> 분, sec -> 초
- document.title useEffect로 관리
- Button type 추가
- api 함수명, 파일명에 http method 추가
- audio 함수명 변경 playSound->play

jungjinbeom committed:
feat: 스터디룸 정보 API 호출  커스텀 훅 추가
- useStudyRoomQuery.ts 커스텀 훅 추가

jungjinbeom committed:
feat: 스터디 룸 정보, Step 정보 API 연동 및 UI 수정
- 스터디룸 조회 API 연동
- Step 정보 업데이트 API 연동
- 서버 데이터와 클라이언트 데이터 동기화
- 스터디룸 조회 parameter타입 수정

jungjinbeom committed:
feat: 스터디 룸 정보, Step 정보 API 연동 및 UI 수정
- 스터디룸 조회 API 연동
- Step 정보 업데이트 API 연동
- 서버 데이터와 클라이언트 데이터 동기화
- 스터디룸 조회 parameter타입 수정

jungjinbeom committed:
chore: 스터디룸 타입 정의 및 수정
- Timer컴포넌트 Props 타입 정의
- useStudyRoomQuery 훅 타입 정의
- getStudyRoomInfo 파라미터 타입 수정

jungjinbeom committed:
Merge branch 'develop' into feat/jjb-timer_create

jungjinbeom committed:
Merge branch 'feat/jjb-timer_create' of https://github.com/junhyeong00/Pomodoro-mate-frontend into feat/jjb-timer_create

jungjinbeom committed:
chore: 타이머 Conflict 수정
- timer 컴포넌트 Props Conflict 수정

jungjinbeom committed:
chore: 타이머 시간 고정
- 타이머 시간 35분으로 고정

jungjinbeom committed:
feat: 스피너 컴포넌트 추가
- API 호출 시 Loading

jungjinbeom committed:
chore: 스터디룸용 레이아웃 적용
- 스터디룸에 레이아웃 컴포넌트 적용

jungjinbeom committed:
chore: 스터디룸 목록,  스터디룸 정보 타입 수정, 중복 파일 및 상수 제거
- 스터디룸 목록 타입, 스터디룸 정보 타입 -> study-room.types.ts에서 관리
- 스터디룸 타입  공통으로 활용 할 수 있게 분리
- 스터리룸 호출 Api Response 타입 수정
- 중복으로 사용되는 Step 파일 및 상수 삭제

jungjinbeom committed:
chore: 사용하지 않는 Step 업데이트 API 및 커스텀 훅 제거
- Step 업데이트 API 제거
- Step 업데이트 API 호출 커스텀 훅 제거
- Step 업데이트 정보 함수 제거
- refetch 함수 제거
